solutions/cf-963D.mdx

```diff
@@ -0,0 +1,175 @@
+---
+id: cf-963D
+source: cf
+title: Frequency of String
+author: Dustin Miao
+---
+
+The [official editorial](https://codeforces.com/blog/entry/58991) mentions a solution using an Aho-Corasick Automaton built over the query strings. We run the initial string through the automaton to locate all occurence positions of each query. The answer can easily be identified for each query by using a sliding window. 
+
+More intuitively, instead of building an Aho-Corasick automaton over the queries, let us build a suffix automaton over the input string. Like before, we will need all occurence positions.
+
+For a state $s$, let $pos[s]$ denote the ending position of the first occurence of any string that corresponds to state $s$. 
+- When we create a new state $s$, then $pos[s] = len[s] - 1$
+- When we clone state $t$ to create a state $s$, then $pos[s] = pos[t]$
+With this method,  we can easily initialize $pos$ without any extra time complexity. Let $occ_P$ be the set of positions where a string $P$ begins in the string. If state $s$ corresponds to $P$, then clearly $(pos[p] - |P| + 1) \in occ_P$. To find the remainder of occurence positions, we can take advantage of the structure of the suffix automaton. 
+
+All states where $P$ is a suffix is a candidate for $occ_P$. Thus, we simply do a BFS/DFS on the _suffix-link tree_, or the tree formed by building a tree out of the suffix links and rooted at the initial state, starting from state that we arrive at when we run $P$ through the suffix automaton. There is a technicality that more than one state may have the same occurence position. Namely, this happens only with cloned states; thus, we redefine $pos[s] = -1$ for all cloned states $s$. 
+
+We can preprocess all queries, and solve them all with a single BFS/DFS. 
+
+<LanguageSection>
+
+<CPPSection>
+
+**Time Complexity:** $\mathcal{O}(m \sqrt{m} \log n)$
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+ 
+using pii = pair<int, int>;
+
+//BeginCodeSnip{Suffix Automaton}
+/**
+ * Generates the Suffix Automaton of a given string
+ * Complexity: O(|S|)
+ */ 
+struct SuffixAuto {
+	struct State {
+		int len, link, pos;
+		int next[26];
+		State(int len = 0, int link = -1, int pos = -1) : len(len), link(link), pos(pos) {
+			memset(next, -1, sizeof(next));
+		}
+	};
+ 
+	vector<State> states;
+	
+	SuffixAuto() {}
+ 
```

Comment:
```suggestion
```
why does this still exist lol

Reply:
you need a constructor...

Reply:
You still need a constructor

Reply:
no he declared one @ the bottom
dustin explained in one of his comments